distance-metrics/src/simple.rs

```diff
@@ -0,0 +1,201 @@
+use super::metric::Metric;
+#[cfg(target_arch = "x86_64")]
+use super::simple_avx::*;
+#[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+use super::simple_neon::*;
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use super::simple_sse::*;
+use super::types::VectorElementType;
+use super::types::ScoreType;
+
+#[cfg(target_arch = "x86_64")]
+const MIN_DIM_SIZE_AVX: usize = 32;
+
+#[cfg(any(
+    target_arch = "x86",
+    target_arch = "x86_64",
+    all(target_arch = "aarch64", target_feature = "neon")
+))]
+const MIN_DIM_SIZE_SIMD: usize = 16;
+
+#[derive(Clone)]
+pub struct DotProductMetric {}
+
+#[derive(Clone)]
+pub struct CosineMetric {}
+
+#[derive(Clone)]
+pub struct EuclidMetric {}
+
+impl Metric for EuclidMetric {
```

Comment:
Style nit: please keep the `impl` block with the type definition. (We usually have the inherent `impl` block first, then trait `impl`s, roughly in order of decreasing rarity.)